enterprise/frontend/src/embedding-sdk/components/public/InteractiveQuestion/context.tsx

```diff
@@ -0,0 +1,171 @@
+import {
+  createContext,
+  type PropsWithChildren,
+  useCallback,
+  useContext,
+  useEffect,
+  useState,
+} from "react";
+import { useUnmount } from "react-use";
+
+import type { SdkPluginsConfig } from "embedding-sdk";
+import { getDefaultVizHeight } from "embedding-sdk/lib/default-height";
+import { useSdkSelector } from "embedding-sdk/store";
+import { getPlugins } from "embedding-sdk/store/selectors";
+import { useDispatch, useSelector } from "metabase/lib/redux";
+import { initializeQBRaw, resetQB } from "metabase/query_builder/actions";
+import {
+  getCard,
+  getFirstQueryResult,
+  getQueryResults,
+  getQuestion,
+  getUiControls,
+} from "metabase/query_builder/selectors";
+import type { Mode } from "metabase/visualizations/click-actions/Mode";
+import { getEmbeddingMode } from "metabase/visualizations/click-actions/lib/modes";
+import type Question from "metabase-lib/v1/Question";
+import type { Card, Dataset } from "metabase-types/api";
+import type { QueryBuilderUIControls } from "metabase-types/store";
+
+type InteractiveQuestionContextType = {
+  question: Question | undefined;
+  card: Card | null;
+  result: Dataset | null;
+  uiControls: QueryBuilderUIControls;
+  queryResults: Dataset[] | null;
+  plugins: SdkPluginsConfig | null;
+  mode: Mode | null | undefined;
+  defaultHeight?: number;
+  isQuestionLoading: boolean;
+  isQueryRunning: boolean;
+  resetQuestion: () => void;
+  onReset?: () => void;
+  onNavigateBack?: () => void;
+  withTitle?: boolean;
+  customTitle?: React.ReactNode;
+  withResetButton?: boolean;
+};
+
+export const InteractiveQuestionContext = createContext<
+  InteractiveQuestionContextType | undefined
+>(undefined);
+
+const returnNull = () => null;
+
+export const InteractiveQuestionProvider = ({
+  children,
+  location,
+  params,
+  componentPlugins,
+  onReset,
+  onNavigateBack,
+  withTitle = false,
+  customTitle,
+  withResetButton,
+}: PropsWithChildren<{
```

Comment:
I wonder if this could be achieved with a hook as opposed to using React context, which requires an extra provider? As most of these are already querying from the Redux selectors so their values would've already been unique.

I think another possible approach would've been to create two hooks, one for loading the question, and another to get the bundled selector values.

```tsx
// Returns a loadQuestion method that can be called
export function useLoadInteractiveQuestion() {
  ...
  return loadQuestion
}

// We may not even need this hook if different components consume different data. Using individual `useSelector` limits the chance of selecting more than you would use.
export function useInteractiveQuestionResult() {
  return ...
}
```

Reply:
For anybody else reading this, I had a conversation with Poom about the future of the SDK in terms of having multiple interactive questions on a page and giving clients more customization on the layout of the interactive question. 

There is an argument for keeping the context, where we can treat the Provider as a sort of "box" that holds the `questionId`, and allows the question components within the provider (i.e. `Visualization`, `Filter`, etc) to share the question data and state across each other without needing to pass props and things like that.

If we look at one of the code snippets in #44730:

```
if (isFilterOpen) {
  return <Filter />;
}

if (isSummarizeOpen) {
  return <Summarize />;
}

return (
  <Box
    className={cx(CS.flexFull, CS.fullWidth)}
    h={height ?? defaultHeight}
    bg="var(--mb-color-bg-question)"
  >
    <Stack h="100%">
      <Flex direction="row" gap="md" px="md" align="center">
        <QuestionBackButton />
        <Title />
        <QuestionResetButton />
        <FilterButton />
        <SummarizeButton />
      </Flex>

      <FilterBar />

      <Group h="100%" pos="relative" align="flex-start">
        <QuestionVisualization />
      </Group>
    </Stack>
  </Box>
);


```

We can see that we can simply put different components within an arbitrary layout, de-coupling the data loading from the actual question view. While we can use a hook for each of these components as mentioned above, we will get to a point in the very near future where we'll have multiple questions on a page and each question's components will need to figure out which `questionId` and context it belongs in. So, at the very least, we'll need to have the `questionId` in some sort of context to keep track of this down the component tree. When we modify the redux store to allow for multiple questions, we'll be able to move some of the logic within the provider to a hook (i.e. `useLoadedQuestion(questionId)` but until then, using a hook will be a temporary solution which will eventually need a context

Reply:
You have a good point about holding `questionId` in the context and I agree in that regard, however I don't think the use of the selectors we have at the moment work with multiple context/providers at the moment. e.g. if we have 2 providers with different `questionId` but inside the provider both will call `useSelector(getQuestion)` which should end up with the same value anyway. Since there can only be one question in the question state at a time. Unless we refactor this to support multiple state, this won't work either.

What I'm seeing that might make more sense here is to extract parts where we use a lot of selectors out into a hook. What this does are:
1. It reduces the complexity of the context, so it only contain necessary data
2. And it should cause less unnecessary rerender (we chatted about this in our DM yesterday). Hooks can be called at any level the same way we get the value from `useContext` anyway, so I think they're both equivalent in terms of reducing prop drilling

Reply:
Added the outcome of our conversation in this [this reply here](https://github.com/metabase/metabase/pull/44730/files#r1657019631)